Private events app

notlfish opened the pull request:
Pull request to add a private events app.
The app should implement:
- User and Event models, with their corresponding information.
- Associations between users and events:
  - A user can host many events.
  - An event has a host user.
  - A user can attend many events.
  - An event can have many guests.
- CRUD actions for events.
- User creation action.
- Authentication and log in.
- Unit and feature tests using rspec and capybara.

notlfish committed:
Generate Event and User models

User model generated using devise.

notlfish committed:
Add creator User->Events association

fortuneonyeka committed:
add user validations

fortuneonyeka committed:
update devise log out and login logic

fortuneonyeka committed:
add event creation functionality

fortuneonyeka committed:
update access to Users#show

notlfish committed:
Add Event#show

notlfish committed:
Add attended_event <-> attendee association.

Has_many through association, using EventAttendance intermediate model.

notlfish committed:
Update User#show to show attended events

fortuneonyeka committed:
seperate events into past and future events

fortuneonyeka committed:
add navigation links for Home and New Event

fortuneonyeka committed:
update past and future events query implementation into scope

fortuneonyeka committed:
update user destroy event

notlfish committed:
Add EventAttendances#destroy

notlfish committed:
Add tests for User and Event models

notlfish committed:
Add signin and signup success tests

notlfish committed:
Clean root folder

fortuneonyeka committed:
update RSpec.describe and rails helper

notlfish committed:
Fix successful sign in tests

notlfish committed:
Finish authentication tests

notlfish committed:
Add event joining test

fortuneonyeka committed:
add The not to assist process test

fortuneonyeka committed:
add The delete process test

fortuneonyeka committed:
update the deleting of the organised event process

fortuneonyeka committed:
update the feature tests

fortuneonyeka committed:
add background color

fortuneonyeka committed:
add css styling

fortuneonyeka committed:
update classes

notlfish committed:
Add styling to Event#index, User#show, and Event#new

notlfish committed:
Add event links underline

fortuneonyeka committed:
add styling to sign in, log out,and sign up forms

fortuneonyeka committed:
remove package-lock file

notlfish committed:
Fix linting errors

fortuneonyeka committed:
add project screenshots

fortuneonyeka committed:
Merge branch 'feature/private-events' of github.com:notlfish/private-events into feature/private-events

notlfish committed:
Update README

Georjane reviewed (changes requested):
### Hello, Team <STUDENTS>, ✋🏽 

Good job so far! You have a very professional Readme file, your PR is professional and you have a good mastery of git-flow. Your linters are all passing, I applaud that too.
However, we have a few more changes to make in order to meet all requirements for this project but you are almost there!

## Required Changes ♻️

- [x] After successful sign in, a user should be redirected to the Event's index page (set to `root_path` in the `Routes` file) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/login_requirement.\*\*png\*\*) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)) and not to create new events as it is in your app
- [x] The join event button throws an error
![image](https://user-images.githubusercontent.com/60772732/124394909-64bce300-dcf9-11eb-8db6-b51d831e52e4.png)
The event's show page displays the list of attendants but  we can't see this because of the error above

#### Controllers

##### EventsController

###### `create` action

* [x] Uses the **#build** [association refrence method](https://guides.rubyonrails.org/association_basics.html#detailed-association-reference), i.e. `current_user.events.build(event_params)` to create the new event with the user’s ID prepopulated.([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

###### `index` action

* [x] Lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] Lists upcoming events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### UsersController

###### `show` action

* [x] Lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] Lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] Lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] No notifications is been shown when a user that doesn't exist tries to sign in


Cheers and Happy coding! 💻 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 🤝 

_Code Reviewer
**[Jane](https://witahgeorjane.tech/)**_

notlfish committed:
Fix errors pointed out in last code review.

Link: https://github.com/notlfish/private-events/pull/2#pullrequestreview-698697013

notlfish committed:
Add live-version link to README

SajjadAhmad14 commented on app/models/event.rb:
```diff
@@ -0,0 +1,21 @@
+  has_many :event_attendances, foreign_key: 'attended_event_id', dependent: :destroy
+  has_many :attendees, through: :event_attendances
+
+  # def self.past_events_of(user)
+  #   user.attended_events.where(["date < ?", DateTime.now])
+  # end
+
+  # def self.future_events_of(user)
+  #   user.attended_events.where(["date > ?", DateTime.now ])
+  # end
```
Please remove commented code from here.

SajjadAhmad14 commented on app/views/devise/mailer/email_changed.html.erb:
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
+  <p>We're contacting you to notify you that your email is being changed to <%= @resource.unconfirmed_email %>.</p>
+<% else %>
```
Please move this logic to app/helpers. [ OPTIONAL ]

SajjadAhmad14 commented on app/views/devise/passwords/edit.html.erb:
```diff
@@ -0,0 +1,25 @@
+
+  <div class="field">
+    <%= f.label :password, "New password" %><br />
+    <% if @minimum_password_length %>
+      <em>(<%= @minimum_password_length %> characters minimum)</em><br />
+    <% end %>
```
Please move this logic to app/helpers. [ OPTIONAL ]

SajjadAhmad14 commented on app/views/devise/shared/_error_messages.html.erb:
```diff
@@ -0,0 +1,15 @@
+<% if resource.errors.any? %>
+  <div id="error_explanation">
+    <h2>
+      <%= I18n.t("errors.messages.not_saved",
+                 count: resource.errors.count,
+                 resource: resource.class.model_name.human.downcase)
+       %>
+    </h2>
+    <ul>
+      <% resource.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+    </ul>
+  </div>
+<% end %>
```
Please move this logic to app/helpers. [ OPTIONAL ]

SajjadAhmad14 commented on app/views/layouts/application.html.erb:
```diff
@@ -1,16 +1,34 @@
+  <div class="navbar">
+    <ul class="nav-list">
+      <li><%=link_to "Home", root_path, class:"home"%></li>
+      <% if user_signed_in?  %>
+      <li><%= link_to 'Log Out', destroy_user_session_path, method: 'delete', class:"log" %></li>
+      <li><%= link_to current_user.name, user_path(current_user), class:"user" %></li>
+      <% else %>
+      <li><%= link_to 'Sign Up', new_user_registration_path, class:"log" %></li>
+      <li><%= link_to 'Log In', new_user_session_path, class:"log" %></li>
+      <% end %>
```
Please move this logic to app/helpers. [ OPTIONAL ]

SajjadAhmad14 reviewed (changes requested):
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, thank you for working on changes suggested by the previous reviewer :ok_hand:
## To Highlight 🎉
- [ ] Git flow is used properly.
- [ ] App is working perfectly.
## To Improve 🙏
- [x] Please remove commented code. It makes the overall codebase unclean.
- [ ] Please add the `npm install` or `yarn install` command in the Setup section and `rspec` to run tests in README.
- [x] Please see inline comments for reference.
# Optional
- [ ] Views should be completely separate from app logic. You can move all view-related logic to app/helpers.
### Happy Coding ☺️

## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

fortuneonyeka committed:
add installation and rspec

fortuneonyeka committed:
remove comment on query class methods

notlfish committed:
Clean up devise-related code.

Erased unused auto-generated files and move logic from views to
app/helpers/application_helper.rb.

notlfish committed:
Merge branch 'experiment' into feature/private-events

Cleanup app/views/devise.

fortuneonyeka committed:
move navbar logics to helpers method

fortuneonyeka committed:
fixed rubocop errors

cyonii reviewed (approved):
## STATUS: APPROVED ✅

You have a really neat project 👏.

Congratulations 🎉, your project has been approved.
You have met the requirements for this project, you can now merge the PR.

> #### <br>Feel free to leave a comment or question if something is not clear. Slack: @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR)<br><br>
